### PR TITLE
fix: Fix StackOverflow in ChangeStreamStateMachine due to excessive mods

### PR DESCRIPTION
**Pull request by tengzhonger**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 1]
```diff
@@ -272,14 +270,13 @@ ChangeStreamStateMachine.State handleCloseStream(
      * DeleteColumn, or a SetCell.
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:270]: This comment also needs to be updated, accepts a new datachange ..

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 2]
```diff
@@ -601,10 +563,10 @@ private State checkAndFinishMutationIfNeeded(
                   dataChange.getEstimatedLowWatermark().getNanos()));
       return AWAITING_STREAM_RECORD_CONSUME;
     }
-    // Case 2_2): The current DataChange itself is chunked, so wait for the next
+    // Case 2: The current DataChange itself is chunked, so wait for the next
     // ReadChangeStreamResponse. Note that we should wait for the new mods instead
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:567]:
```suggestion
    // ReadChangeStreamResponse. Note that we should wait for the new data change instead
```

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 3]
```diff
@@ -356,161 +351,138 @@ State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
...
    * <dl>
    *   <dt>Valid exit states:
-   *   <dd>{@link ChangeStreamStateMachine#AWAITING_NEW_MOD}. Current mod is added, and we have more
-   *       mods to expect.
-   *   <dd>{@link ChangeStreamStateMachine#AWAITING_CELL_VALUE}. Current mod is the first chunk of a
-   *       chunked SetCell.
-   *   <dd>{@link ChangeStreamStateMachine#AWAITING_STREAM_RECORD_CONSUME}. Current mod is the last
-   *       mod of the current logical mutation.
+   *   <dd>{@link ChangeStreamStateMachine#AWAITING_NEW_DATA_CHANGE}. All mods from the current
+   *       DataChange are added, and we have more mods to expect.
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:364]:
```suggestion
   *       DataChange are added, and we have more DataChange to expect.
```

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 4]
```diff
@@ -356,161 +351,138 @@ State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
...
-              validate(
-                  chunk.getChunkInfo().getChunkedValueOffset() == 0,
-                  "AWAITING_NEW_MOD: First chunk of a chunked cell must start with offset==0.");
-              validate(
-                  chunk.getChunkInfo().getChunkedValueSize() > 0,
-                  "AWAITING_NEW_MOD: First chunk of a chunked cell must have a positive chunked value size.");
-              expectedTotalSizeOfChunkedSetCell = chunk.getChunkInfo().getChunkedValueSize();
-              actualTotalSizeOfChunkedSetCell = 0;
+        State handleDataChange(ReadChangeStreamResponse.DataChange dataChange) {
+          // Iterate over all mods.
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:385]: Did you explain what a mod is anywhere? Maybe explain a little bit here:

there are 3 types of mod:
setCell,
deleteFamily,
deleteFromColumn.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 5]
```diff
@@ -356,161 +351,138 @@ State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
...
+                          + "; actual total size: "
+                          + actualTotalSizeOfChunkedSetCell);
+                  continue;
+                } else {
+                  // If this is not the last chunk of a chunked SetCell, then this must be the last
+                  // mod of the current response, and we're expecting the rest of the chunked cells
+                  // in the following ReadChangeStream response.
+                  validate(
+                      index == dataChange.getChunksCount() - 1,
+                      "AWAITING_CELL_VALUE: Current mod is a chunked SetCell "
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java]:
```suggestion
                      "AWAITING_NEW_DATA_CHANGE: Current mod is a chunked SetCell "
```

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 6]
```diff
@@ -356,161 +351,138 @@ State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
...
+                } else {
+                  // Case 1_2_2
+                  validate(
+                      index == 0,
+                      "AWAITING_NEW_DATA_CHANGE: Non-first chunked SetCell must be the first mod of a DataChange.");
+                }
+                // Concatenate the cell value of this mod into the builder.
+                validate(
+                    chunk.getChunkInfo().getChunkedValueSize() == expectedTotalSizeOfChunkedSetCell,
+                    "AWAITING_CELL_VALUE: Chunked value size must be the same for all chunks.");
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java]:
```suggestion
                    "AWAITING_NEW_DATA_CHANGE: Chunked cell value size must be the same for all chunks.");
```

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 7]
```diff
@@ -356,161 +351,138 @@ State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
...
+                numCellChunks++;
+                builder.cellValue(setCell.getValue());
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:438]: Question, will this always concatenate cell values?

**Review by mutianf — APPROVED** [review]
lgtm with some nits

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java (hunk 4 above)
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:385]: Yeah I've updated the comments of `ChangeStreamStateMachine.State handleDataChange` to say that `A DataChange can have multiple mods, where each mod could be a DeleteFamily, a DeleteColumn, or a SetCell`

**Review by mutianf — APPROVED** [review]

**Pull request merged**